src/boot.s

```diff
@@ -34,6 +34,15 @@ _start:
 single_core:                            # only the 1st hart past this point
         la      sp, stack_top           # setup stack pointer
 
+        la      a0, bss_start
+        la      a1, bss_end
+        bgeu    a0, a1, init
+clean_bss_loop:
+        sd      zero, (a0)
```

Comment:
There's a subtle (and obscure, on our side) problem here: we build both 64- and 32-bit targets, so using `sd` instruction directly fails to build for 32 bits:

```
$ make all
[...]
riscv64-linux-gnu-gcc -march=rv32g -mabi=ilp32 -static -mcmodel=medany -fvisibility=hidden -nostdlib -nostartfiles -ffreestanding -Tsrc/baremetal.ld -Iinclude \
	-Wl,--defsym,ROM_START=0x20400000 -Wa,--defsym,UART=0x10013000 \
	-Wa,--defsym,XLEN=32 \
	src/baremetal-fib.s src/baremetal-print.s src/baremetal-poweroff.s -o out/test_sifive_e32
riscv64-linux-gnu-gcc -march=rv32g -mabi=ilp32 -static -mcmodel=medany -fvisibility=hidden -nostdlib -nostartfiles -ffreestanding -Tsrc/baremetal.ld -Iinclude \
	-Wl,--defsym,ROM_START=0x20400000 -Wa,--defsym,UART=0x10013000 \
	-Wa,--defsym,XLEN=32 \
	src/boot.s src/baremetal-print.s src/baremetal-poweroff.s src/userland.c src/kernel.c src/syscalls.c src/pmp.c src/riscv.c src/fdt.c src/string.c src/proc_test.c -o out/user_sifive_e32
src/boot.s: Assembler messages:
src/boot.s:41: Error: unrecognized opcode `sd zero,(a0)'
Makefile:163: recipe for target 'out/user_sifive_e32' failed
make: *** [out/user_sifive_e32] Error 1
```

We have a macro `sx` defined for that [here](https://github.com/rtfb/riscv64-in-qemu/blob/master/src/machine-word.inc#L25-L27). You can find a usage example [here](https://github.com/rtfb/riscv64-in-qemu/blob/master/src/boot.s#L138).

(And I wonder why github actions didn't catch this... This is something I'll have to look into).

Reply:
Okay, looks like this is the reason why github actions didn't run: https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks

#TIL